tsnet/tsnet.go

```diff
@@ -997,10 +1001,11 @@ type listenKey struct {
 }
 
 type listener struct {
-	s    *Server
-	keys []listenKey
-	addr string
-	conn chan net.Conn
+	s      *Server
+	keys   []listenKey
+	addr   string
+	conn   chan net.Conn
+	closed bool // guarded by s.mu
```

Comment:
I think all of the fields are guarded by s.mu, it isn't safe to manipulate s.listeners without holding s.mu.

Reply:
These are immutable after creation, so they can be read without holding a lock.